R/get_tiles.R

```diff
@@ -0,0 +1,155 @@
+  # Error handling
+  # Collect input arguments as a list
+  args <- as.list(environment())
+  # Check whether class of arguments is valid
+  ref <- list(
+    zoom = "integer",
+    x = "integer",
+    y = "integer",
+    scale = "character"
+  )
+  check_arguments(x = args, ref = ref)
+  # if x or y is NULL, get all tiles in that dimension
+  if (is.null(x)) x <- 0:(2 ^ zoom - 1)
+  if (is.null(y)) y <- 0:(2 ^ zoom - 1)
+  # Get specified tiles
+  tiles_list <- list()
+  # TODO: add a progress bar
+  for (i in x) {
+    for (j in y) {
+      # Check if the tile exists
+      url <- paste0(
+        "https://tiles.macrostrat.org/", scale, "/",
+        zoom, "/", i, "/", j, ".mvt"
+      )
+      if (http_error(url)) {
+        stop("Tile does not exist: ", url)
+      }
+      tiles_list[[paste0(i, "/", j)]] <- read_mvt_sf(url)
+    }
+  }
+  # remove tiles without data
+  tiles_filt <- Filter(length, tiles_list)
+  # combine the tiles
+  tiles <- list()
+  # TODO: use st_union to combine duplicates based on map_id
+  lines <- lapply(tiles_filt, function(x) x$lines)
+  lines_filt <- Filter(Negate(is.null), lines)
+  tiles$lines <- st_rbindlist(lines_filt, use.names = TRUE, ignore.attr = TRUE,
+                              fill = TRUE, idcol = "tile")
+  units <- lapply(tiles_filt, function(x) x$units)
+  units_filt <- Filter(Negate(is.null), units)
+  tiles$units <- st_rbindlist(units_filt, use.names = TRUE, ignore.attr = TRUE,
+                              fill = TRUE, idcol = "tile")
+  return(tiles)
+  # TODO: document variables in return object
+}
+
+# MODIFIED FROM https://github.com/a-benini/sfhelpers
+#' @importFrom data.table rbindlist
+#' @importFrom sf st_as_sf st_crs st_geometry_type
+#' @importFrom uuid UUIDgenerate
+st_rbindlist <- function(l, ..., geometry_name = NULL) {
+  if (!is.list(l) || is.data.frame(l)) {
+    stop("input is ", class(l)[1L], " but should be a plain list of sf objects",
+         " to be stacked", call. = FALSE)
+  }
+  if (!(is.null(geometry_name) | (is.character(geometry_name) &
+                                 length(geometry_name) == 1))) {
+    stop("geometry_name must be either NULL or a single character string",
+         call. = FALSE)
+  }
+  is_not_null <- !vapply(l, is.null, logical(1))
+  if (!all(vapply(l[is_not_null], inherits, what = "sf", logical(1)))) {
+    stop("not all listed objects are of the class sf", call. = FALSE)
+  }
+  if(!any(is_not_null)) {
+    stop("no sf objects included in input list", call. = FALSE)
+  }
+  if (length(unique(lapply(l[is_not_null], st_crs))) > 1) {
+    stop("arguments have different crs", call. = FALSE)
+  }
+  # name of geometry column of the first listed sf-object
+  first_geometry_name <- attr(l[is_not_null][[1]], "sf_column")
+
+  # homogenize name of active geometry columns
+  if (length(unique(vapply(l[is_not_null], attr,
+                           which = "sf_column", character(1)))) > 1) {
+    tmp_geometry_name <- UUIDgenerate()
```

Comment:
Strictly needed in our case? Just thinking about reducing the number of packages we're importing from